### ISSUE TITLE
How to customize httpd for PSGI application

### ISSUE THREAD

**Issue by ppisar** [openshift/sti-perl#58]
PSGI applications do not seem to work with ModPerl::PerlRun PerlResponseHandler.

This simple application:

~~~~
$ cat index.cgi 
#!/usr/bin/env perl
use Plack::Runner;
use Web::Paste::Simple;
Plack::Runner->new->run(Web::Paste::Simple->new->app);
$ cat cpanfile 
requires 'Plack::Runner';
requires 'Web::Paste::Simple';
~~~~

sends empty response on subsequent requests. (Actually after all preforked httpd workers are exhausted and are recycled by httpd.) I don't know if this is a bug in mod_perl, Plack PSGI server, or the Web::Paste::Simple application.

But any Plack documentation states the right way is to use Plack::Handler::Apache2 PerlResponseHandler instead of the ModPerl::PerlRun one (http://search.cpan.org/dist/Plack/lib/Plack/Handler/Apache2.pm). And that indeed works. (Another option is to use Plack::Handler::Apache2::Registry PerlResponseHandler).

So we need to be able to redefine the PerlResponseHandler in the httpd configuration.

Also users may want have nice short URLs without the visible real application file name (e.g. http://domain/ instead of http://domain/application.psgi). Thus the recommended way is to adjust httpd configuration for every application like:

~~~~
<Location />
SetHandler perl-script
PerlResponseHandler Plack::Handler::Apache2
PerlSetVar psgi_app /path/to/application.psgi
</Location>
~~~~

This requires to know that the deployed code is an PSGI application and to know where the file (application.psgi) is installed by absolute path (/path/to/application.psgi).

My question is how to augment the assemble script to support this use case?

We could:

(1) Allow user to supply httpd configuration file in the sources.

(2) Allow user to specify path to the PSGI file in the source tree with an environment variable and generate the httpd configuration based on the varaible.

(3) Scan the sources for PSGI files or expect some hard coded file name and generate a the httpd configuration.

Is there a preferred method? How are similar issues handled by other httpd docker images?

**Comment by ppisar** [comment]
After discussion with people maintaining httpd, mod_perl and Ruby image, I think the best approach is to scan top-level source directory for exactly one '*.psgi' file and automatically redirect "/" location to the file by amending httpd's configuration. This will allow to deploy PSGI applications easily.

For those with specific needs, there will be an environment variable to select the PSGI file explicitly. Maybe there will be another variable to define the location explicitly (instead of "/") to allow httpd serving other files by normal means.

**Comment by ppisar** [comment]
Implemented since f5a7e2de96fa2a318ea09b2c428feb9c24be3563.
